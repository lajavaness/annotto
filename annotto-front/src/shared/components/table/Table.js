import React, { useCallback, useEffect, useRef, useState } from 'react'

import Loader from 'shared/components/loader/Loader'

import * as Styled from './__styles__/Table.styles'

const Table = (props) => {
	const rootRef = useRef(null)

	const [height, setHeight] = useState(0)
	const [headerHeight, setHeaderHeight] = useState(0)
	const [footerHeight, setFooterHeight] = useState(0)
	const [isResized, setIsResized] = useState(false)

	const _onWindowResize = useCallback(() => setIsResized(false), [])

	const getHeight = (node) => (node ? node.getBoundingClientRect().height : 0)

	useEffect(() => {
		_onWindowResize()
		window.addEventListener('resize', _onWindowResize)
		return () => window.removeEventListener('resize', _onWindowResize)
	}, [_onWindowResize])

	useEffect(() => {
		if (rootRef?.current) {
			if (!isResized) {
				setHeight(getHeight(rootRef.current))
				setIsResized(true)
			} else {
				const header = rootRef.current.querySelector('thead.ant-table-thead')
				const footer = rootRef.current.querySelector('div.ant-table-footer')
				setHeaderHeight(getHeight(header))
				setFooterHeight(getHeight(footer))
			}
		}
	}, [rootRef, isResized, height])

	return (
		<Styled.Root ref={rootRef}>
			{isResized ? (
				<Styled.Table
					size="middle"
					tableLayout="auto"
					scroll={{ x: 'max-content', y: headerHeight ? height - headerHeight - footerHeight : 0 }}
					pagination={false}
					{...props}
				/>
			) : (
				<Loader />
			)}
		</Styled.Root>
	)
}

export default Table
