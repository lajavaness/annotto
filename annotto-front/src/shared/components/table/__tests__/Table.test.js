import { ThemeProvider } from 'styled-components'
import { render } from '@testing-library/react'
import React from 'react'

import theme from '__theme__'

import Table from 'shared/components/table/Table'

const getInstance = () => (
	<ThemeProvider theme={theme}>
		<Table />
	</ThemeProvider>
)

describe('Table', () => {
	it('matches snapshot', () => {
		const { asFragment } = render(getInstance())
		expect(asFragment()).toMatchSnapshot()
	})
})
