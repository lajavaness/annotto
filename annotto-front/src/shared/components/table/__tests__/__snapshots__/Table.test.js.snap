// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-aKXVD fEmcyS"
  >
    <div
      class="ant-table-wrapper sc-kmATbt AOMwP"
    >
      <div
        class="ant-spin-nested-loading"
      >
        <div
          class="ant-spin-container"
        >
          <div
            class="ant-table ant-table-middle ant-table-empty ant-table-fixed-header ant-table-scroll-horizontal"
          >
            <div
              class="ant-table-container"
            >
              <div
                class="ant-table-header"
                style="overflow: hidden;"
              >
                <table
                  style="table-layout: fixed;"
                >
                  <thead
                    class="ant-table-thead"
                  >
                    <tr>
                      <th
                        class="ant-table-cell"
                      />
                    </tr>
                  </thead>
                </table>
              </div>
              <div
                class="ant-table-body"
                style="overflow-x: auto; overflow-y: scroll; max-height: 0;"
              >
                <table
                  style="min-width: 100%; table-layout: auto;"
                >
                  <colgroup />
                  <tbody
                    class="ant-table-tbody"
                  >
                    <tr
                      aria-hidden="true"
                      class="ant-table-measure-row"
                      style="height: 0px; font-size: 0px;"
                    >
                      <td
                        style="padding: 0px; border: 0px; height: 0px;"
                      >
                        <div
                          style="height: 0px; overflow: hidden;"
                        >
                          Â 
                        </div>
                      </td>
                    </tr>
                    <tr
                      class="ant-table-placeholder"
                    >
                      <td
                        class="ant-table-cell"
                      >
                        <div
                          class="ant-table-expanded-row-fixed"
                          style="width: 0px; position: sticky; left: 0px; overflow: hidden;"
                        >
                          <div
                            class="ant-empty ant-empty-normal"
                          >
                            <div
                              class="ant-empty-image"
                            >
                              <svg
                                class="ant-empty-img-simple"
                                height="41"
                                viewBox="0 0 64 41"
                                width="64"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <g
                                  fill="none"
                                  fill-rule="evenodd"
                                  transform="translate(0 1)"
                                >
                                  <ellipse
                                    class="ant-empty-img-simple-ellipse"
                                    cx="32"
                                    cy="33"
                                    rx="32"
                                    ry="7"
                                  />
                                  <g
                                    class="ant-empty-img-simple-g"
                                    fill-rule="nonzero"
                                  >
                                    <path
                                      d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                                    />
                                    <path
                                      class="ant-empty-img-simple-path"
                                      d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                                    />
                                  </g>
                                </g>
                              </svg>
                            </div>
                            <div
                              class="ant-empty-description"
                            >
                              No Data
                            </div>
                          </div>
                        </div>
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
